hypershift-operator/controllers/machineconfigserver/machineconfigserver_controller.go

```diff
@@ -187,44 +194,86 @@ func (r *MachineConfigServerReconciler) Reconcile(ctx context.Context, req ctrl.
 	}
 
 	r.Log.Info("Creating userdata secret")
-	semversion, err := semver.Parse(releaseImage.Version())
+	mcs.Status.Version = releaseImage.Version()
+	reconcileResult, err = r.reconcileUserData(ctx, mcs)
+	if reconcileResult != nil {
+		return *reconcileResult, err
+	} else if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.Status().Update(ctx, mcs); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MachineConfigServerReconciler) reconcileUserData(ctx context.Context, mcs *hyperv1.MachineConfigServer) (*ctrl.Result, error) {
+	semversion, err := semver.Parse(mcs.Status.Version)
 	if err != nil {
-		return ctrl.Result{}, nil
+		return &ctrl.Result{}, err
 	}
-	userDataSecret = MachineConfigServerUserDataSecret(mcs)
-	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, userDataSecret, func() error {
-		// For now, only create and never update this secret
-		if !userDataSecret.CreationTimestamp.IsZero() {
-			return nil
+	userDataSecret := MachineConfigServerUserDataSecret(mcs)
+	r.Log.Info("Gathering data to generate machine userdata secret")
+	var nodeBootstrapperSecret corev1.Secret
+	if err := r.Client.Get(ctx, ctrlclient.ObjectKey{Namespace: mcs.Namespace, Name: haproxyConfigSecretName}, &nodeBootstrapperSecret); err != nil {
+		return &ctrl.Result{}, fmt.Errorf("failed to get machine config server haproxy secret conf %s: %w", haproxyConfigSecretName, err)
+	}
+	var nodeBootstrapperTokenData []byte
+	var ok bool
+	if nodeBootstrapperTokenData, ok = nodeBootstrapperSecret.Data["node-bootstrapper-token"]; !ok {
+		return &ctrl.Result{}, fmt.Errorf("could not find node bootstrapper token in machine config server haproxy secret conf  %s", haproxyConfigSecretName)
+	}
+	var combinedCA string
+	switch mcs.Spec.IgnitionService.Type {
+	case hyperv1.NodePort:
+		r.Log.Info("Gathering hosted cluster CA")
+		var hostedClusterConfigOperatorConfigMapData corev1.ConfigMap
+		if err := r.Client.Get(ctx, ctrlclient.ObjectKey{Namespace: mcs.Namespace, Name: hostedClusterConfigOperatorConfigMapName}, &hostedClusterConfigOperatorConfigMapData); err != nil {
+			return &ctrl.Result{}, fmt.Errorf("failed to get hosted cluster config operator configmap %s: %w", hostedClusterConfigOperatorConfigMapName, err)
+		}
+		if combinedCA, ok = hostedClusterConfigOperatorConfigMapData.Data["initial-ca.crt"]; !ok {
+			return &ctrl.Result{}, fmt.Errorf("could not find node initial-ca.crt in configmap %s", hostedClusterConfigOperatorConfigMapName)
+		}
+	case hyperv1.Route:
+		r.Log.Info("Gathering default router CA")
```

Comment:
Assumption: default router CA will TLS validate the ignition route (tangibly means that ignition route exposed through default router)